File: Interpreter.java
/*****
 * Interpreter
 * Define an interpreter to interpret sentences in a
 * language based on grammatical representations.
*****/

import java.util.Map;
import java.util.HashMap;


interface Expression { int interpret(Map<String, Integer> context); }

class VariableExpression implements Expression {
    private final String variable;
    
    public VariableExpression(String variable) { this.variable = variable; }
    
    @Override
    public int interpret(Map<String, Integer> context) { return context.get(variable); }
}


public class Interpreter {
    public static void main(String[] args) {
        Expression v = new VariableExpression("x");
        Map<String, Integer> context = new HashMap<>();
        context.put("x", 10);
        context.put("y", 20);
        
        int result = v.interpret(context);
        System.out.println(result);
    }
}

File: Interpreter.py
"""
Interpreter
Define an interpreter to interpret sentences in a
language based on grammatical representations.
"""

from abc import ABC, abstractmethod


class Expression(ABC):
    @abstractmethod
    def interpret(self, context):
        pass

class VariableExpression(Expression):
    def __init__(self, variable):
        self.variable = variable
    
    def interpret(self, context):
        if self.variable not in context:
            raise ValueError(f"Variable {self.variable} not found in context")
        return context[self.variable]


if __name__ == "__main__":
    v = VariableExpression('x')
    context = {'x': 10, 'y': 20}
    result = v.interpret(context)
    
    print(result)
